JShellAPI/src/main/java/org/togetherjava/jshellapi/service/DockerService.java

```diff
@@ -62,48 +75,147 @@ private void cleanupLeftovers(UUID currentId) {
+    }
 
-        if (!presentLocally) {
-            client.pullImageCmd(jshellWrapperBaseImageName)
-                .withTag("master")
-                .exec(new PullImageResultCallback())
-                .awaitCompletion(5, TimeUnit.MINUTES);
-        }
+    /**
+     * Pulls the Docker image.
+     */
+    private void pullImage() throws InterruptedException {
+        client.pullImageCmd(jshellWrapperBaseImageName)
+            .withTag("master")
+            .exec(new PullImageResultCallback())
+            .awaitCompletion(5, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Creates a Docker container with the given name.
+     *
+     * @param name The name of the container to create.
+     * @return The ID of the created container.
+     */
+    private String createContainer(String name) {
+        HostConfig hostConfig = HostConfig.newHostConfig()
+            .withAutoRemove(true)
+            .withInit(true)
+            .withCapDrop(Capability.ALL)
+            .withNetworkMode("none")
+            .withPidsLimit(2000L)
+            .withReadonlyRootfs(true)
+            .withMemory((long) config.dockerMaxRamMegaBytes() * 1024 * 1024)
+            .withCpuCount((long) Math.ceil(config.dockerCPUsUsage()))
+            .withCpusetCpus(config.dockerCPUSetCPUs());
 
         return client
             .createContainerCmd(jshellWrapperBaseImageName + Config.JSHELL_WRAPPER_IMAGE_NAME_TAG)
-            .withHostConfig(HostConfig.newHostConfig()
-                .withAutoRemove(true)
-                .withInit(true)
-                .withCapDrop(Capability.ALL)
-                .withNetworkMode("none")
-                .withPidsLimit(2000L)
-                .withReadonlyRootfs(true)
-                .withMemory(maxMemoryMegs * 1024 * 1024)
-                .withCpuCount(cpus)
-                .withCpusetCpus(cpuSetCpus))
+            .withHostConfig(hostConfig)
             .withStdinOpen(true)
             .withAttachStdin(true)
             .withAttachStderr(true)
             .withAttachStdout(true)
-            .withEnv("evalTimeoutSeconds=" + evalTimeout.toSeconds(),
-                    "sysOutCharLimit=" + sysoutLimit)
+            .withEnv("evalTimeoutSeconds=" + config.evalTimeoutSeconds(),
+                    "sysOutCharLimit=" + config.sysOutCharLimit())
             .withLabels(Map.of(WORKER_LABEL, WORKER_UNIQUE_ID.toString()))
             .withName(name)
             .exec()
             .getId();
     }
 
-    public InputStream startAndAttachToContainer(String containerId, InputStream stdin)
+    /**
+     * Spawns a new Docker container with specified configurations.
+     *
+     * @param name Name of the container.
+     * @param startupScriptId Script to initialize the container with.
+     * @return The ContainerState of the newly created container.
+     */
+    public ContainerState initializeContainer(String name, StartupScriptId startupScriptId)
+            throws IOException {
```

Comment:
Log INFO when this method is called, the log should contain the values of the parameters.